Support `INSERT ... INTO ... SELECT ...` for Postgres table

YuweiXiao opened the pull request:
Partially fixes #30

Commit:
initial implement for heap table insert_select

handle type conversion

handle column mapping & cleanup

fix schedule

code cleanup & add restriction

revert uncessary changes

JelteF commented on src/pgduckdb_planner.cpp:
```diff
@@ -33,7 +33,23 @@ extern "C" {
-	const char *query_string = pgduckdb_get_querydef(copied_query);
+	const char *query_string;
+	if (copied_query->commandType == CMD_INSERT) {
+		ListCell *l;
+		RangeTblEntry *select_rte = NULL;
+		foreach(l, copied_query->rtable) {
+			RangeTblEntry *rte = (RangeTblEntry *)lfirst(l);
```
```suggestion
		RangeTblEntry *select_rte = NULL;
		foreach_node(RangeTblEntry, rte, copied_query->rtable) {
```

JelteF commented on src/pgduckdb_hooks.cpp:
```diff
@@ -169,10 +182,16 @@ IsAllowedStatement(Query *query, bool throw_error) {
 		RangeTblEntry *resultRte = list_nth_node(RangeTblEntry, query->rtable, query->resultRelation - 1);
-		if (!::IsDuckdbTable(resultRte->relid)) {
+		if (::IsHeapTable(resultRte->relid) && query->commandType == CMD_INSERT) {
+			if (query->onConflict || query->returningList) {
+				elog(elevel, "DuckDB does not support INSERT ON CONFLICT or RETURNING on Postgres tables");
+				return false;
+			}
```
Why not allow these? Do they not work? I would expect that if we insert using postgres its functionality that thees things would automatically work.

JelteF replied:
To be clear, not saying that this should be implemented in the first version of this feature. But I'd at least like a code comment explaining why we do not support it.

YuweiXiao replied:
Ah, block them since I haven't tested them yet. I've done some initial testing, and RETURNING works out of box while onConflict doesn't. Will dig deeper into the WHY and add comments to clarify.

YuweiXiao replied:
ON CONFLICT is also working

JelteF commented on src/pgduckdb_hooks.cpp:
```diff
@@ -169,10 +182,16 @@ IsAllowedStatement(Query *query, bool throw_error) {
 	if (query->commandType != CMD_SELECT) {
 		RangeTblEntry *resultRte = list_nth_node(RangeTblEntry, query->rtable, query->resultRelation - 1);
-		if (!::IsDuckdbTable(resultRte->relid)) {
+		if (::IsHeapTable(resultRte->relid) && query->commandType == CMD_INSERT) {
```
Why only heap tables? I'd expect this should logic would work fine for most table access method, because they implement the table access method interface.

YuweiXiao replied:
YES, you are right! This is becoming much more exciting if we get the whole thing functional.

JelteF reviewed (changes requested):
CI is failing in many places, because `duckdb.force_execution` is `true` in our tests. Apparently something breaks when selecting from `generate_series()` in duckdb for a PG INSERT.

YuweiXiao commented:
> I think that would be worked around by #583, but then the opposite would hold (i.e. inserting into a varchar column woudl fail). So I think we need some logic to make sure that the types match against eachother, and maybe add some code to convert the type that we get back from DuckDB to the type that PG wants for the INSERT. I think that would also fix the `generate_series()` issue that all of the tests are failing on currently.

YES, the type check and conversion are substantial. I manually added an `Expr` to handle the coerce and it seems working so far (e.g., `varchar` and `generate_series` issue). I'll add more test cases to ensure if I haven't missed (or broken) anything, and code re-org (& cleanup) are also in progress. Thanks for the feedback, @JelteF !

YuweiXiao commented:
The MR is ready for review~ @JelteF Please take a look when you have a chance.

Currently, the regression is failing for `hugeint_numeric`. The issue seems related to the following behavior, which is reproducible in the `main` branch. Do you have any insights on this?

```
select * from duckdb.query($$ SELECT a FROM (VALUES (32942348563242.111222333444555666777888::numeric(38,24)), (NULL::numeric), (123456789.000000000000000000000001::numeric(38,24))) t(a) $$);

                   a
----------------------------------------
 32942348563242.11122233344455566677789  (correct one should be `32942348563242.111222333444555666777888`

      123456789.00000000000000000000000 (correct one should be 123456789.000000000000000000000001)
(3 rows)
```

JelteF commented on include/pgduckdb/pgduckdb_planner.hpp:
```diff
@@ -14,3 +14,4 @@ extern bool duckdb_explain_ctas;
 PlannedStmt *DuckdbPlanNode(Query *parse, const char *query_string, int cursor_options, ParamListInfo bound_params,
                             bool throw_error);
 duckdb::unique_ptr<duckdb::PreparedStatement> DuckdbPrepare(const Query *query, const char *explain_prefix = NULL);
+bool IsAllowedPostgresInsert(Query *query, bool throw_error = false);
```
Seems like you don't have https://editorconfig.org/ installed in your editor. Could you do that and re-save this file to add a trailing newline.

YuweiXiao replied:
sure

JelteF commented on src/pgduckdb_planner.cpp:
```diff
@@ -6,34 +6,107 @@
+		return false;
+	}
+	
+	/* TODO: Checking supported target column types */
```
You wanna implement this TODO still?

YuweiXiao replied:
ah, the TODO comment can be removed. The follow-up code handles the checking through blacklist

JelteF commented:
> Currently, the regression is failing for hugeint_numeric. The issue seems related to the following behavior, which is reproducible in the main branch. Do you have any insights on this?

I think you can probably fix the test by changing the plain `NULL` to `NULL::NUMERIC(38, 24)`. It seems at least that that solves the issue when running your reproducing example in the duckdb CLI.

JelteF commented on src/pgduckdb_planner.cpp:
```diff
@@ -6,34 +6,107 @@
+			continue;
+		}
+
+		if (attr->atttypid == BYTEAOID || attr->atttypid == XMLOID || pgduckdb::pg::IsDomainType(attr->atttypid) ||
```
Instead of hardcoding some unsupported types, I think we should instead determine the allowed types. Also, why are these types unsupported? Because we do support BYTEA as a type for regular duckdb execution, so that seems surprising that it doesn't work here.

YuweiXiao replied:
Changed the behavior to whitelist. As long as the Postgres type can be converted to a DuckDB type using `ConvertPostgresToDuckColumnType`, it will be allowed.

JelteF commented on test/regression/expected/postgres_insert_select.out:
```diff
@@ -0,0 +1,103 @@
+INSERT INTO tbl select r['a']::int, r['b'] from duckdb.query($$ SELECT 21 a, 'ghi' b $$) r;
+ERROR:  no partition of relation "tbl" found for row
+DETAIL:  Partition key of the failing row contains (a) = (21).
+SELECT * FROM tbl ORDER BY a LIMIT 5;
```
Would be good to also check the contents of each partition, to double check that values go to the correct partition.

YuweiXiao replied:
Fixed

JelteF commented on src/pgduckdb_planner.cpp:
```diff
@@ -6,34 +6,107 @@
+	}
+
+	if (!select_rte) {
+		elog(elevel, "DuckDB does not support INSERT without a subquery");
```
There are no tests for this error case afaict

YuweiXiao replied:
Added

JelteF commented on src/pgduckdb_planner.cpp:
```diff
@@ -6,34 +6,107 @@
+IsAllowedPostgresInsert(Query *query, bool throw_error) {
+	int elevel = throw_error ? ERROR : DEBUG4;
+	if (query->commandType != CMD_INSERT) {
+		elog(elevel, "DuckDB only supports INSERT/SELECT on Postgres tables");
```
There are no tests for this error case afaict

YuweiXiao replied:
Added

JelteF commented on test/regression/expected/postgres_insert_select.out:
```diff
@@ -0,0 +1,103 @@
+
+DROP TABLE tbl;
+-- case: ON CONFLICT DO NOTHING
+CREATE TABLE tbl (a int PRIMARY KEY, b text);
```
What happens with a `SERIAL` column or a column with `GENERATED ALWAYS AS IDENTITY`.

JelteF commented:
To set expectations: I think this is a very cool feature and your implementation looks like the right direction. But it is relatively complex and impactful, so I would like to give it an appropriate round of testing and "baking" (aka testing by merging it and having it on the `main` branch for a while). Since we want to release 1.0 somewhat soon, I don't think it's realistic to still do that before the 1.0 release. So I added this to the new 1.1 milestone instead. I also won't spend time reviewing this PR in depth until then.

YuweiXiao commented:
> To set expectations: I think this is a very cool feature and your implementation looks like the right direction. But it is relatively complex and impactful, so I would like to give it an appropriate round of testing and "baking" (aka testing by merging it and having it on the `main` branch for a while). Since we want to release 1.0 somewhat soon, I don't think it's realistic to still do that before the 1.0 release. So I added this to the new 1.1 milestone instead. I also won't spend time reviewing this PR in depth until then.

Totally agree, 1.1 makes sense. btw, any target date for 1.0 release?

JelteF commented:
> any target date for 1.0 release?

The current plan is to do it quickly after the DuckDB 1.3.0 release, so early May. But dates may obviously slip.

Commit:
more regress

Commit:
more regress

Commit:
format

Commit:
fix regress

JelteF committed:
Merge branch 'main' into heap_insert_select

Since this branch was created `DuckdbPlanNode` was rewritten on main to
build a fresh `PlannedStmt` instead of overwriting the plan produced by
`standard_planner`. That made the approach of this branch, which hangs the
DuckDB `CustomScan` under the `ModifyTable` node of the `standard_planner`
plan, no longer apply cleanly. The `IsAllowedPostgresInsert` case now
explicitly calls `standard_planner` to build the INSERT plan, grafts the
DuckDB scan into it, and appends the DuckDB RTE to its rtable (including
the `custom_scan_tlist` varno fixup that main removed for regular
queries). All other queries use main's new from-scratch `PlannedStmt`.

Some other adjustments that were needed to make this branch work on top of
main:

- `pgduckdb::IsDuckdbTable(Oid)` is now implemented using main's
  `DuckdbTableAmGetName`, so that it recognizes every registered DuckDB
  table access method instead of only the built-in one. The static
  `IsDuckdbTable` helpers in the hooks file were replaced by it.
- Unsupported column types are detected with `LogicalTypeId::INVALID`,
  since the newer DuckDB version removed `LogicalTypeId::USER` which
  `IsAllowedPostgresInsert` used for this.
- The `postgres_insert_select` test needed explicit `::text` casts in the
  partitioned table statements, because assigning `duckdb.unresolved_type`
  to a column is no longer allowed on main.
- The expected output of `query_filter` and `parallel_postgres_scan` now
  contains NOTICE/WARNING messages, because with this feature INSERTs into
  Postgres tables attempt DuckDB execution (falling back to Postgres) in
  situations where they previously weren't considered for it at all.

JelteF committed:
Fix column mapping of INSERT ... SELECT into Postgres tables

`ReconstructTargetListForInsert` assigned the columns of the DuckDB result
to the columns of the target table sequentially. That's wrong when the
INSERT lists the columns in a different order than the table definition:
`INSERT INTO tbl (b, a) SELECT ...` would insert the first SELECT column
into `a` instead of `b`, silently inserting wrong data (or erroring when
the types happened to be incompatible). It was also only invoked when the
column counts differed, so a reordered INSERT that assigns *all* columns
skipped the reconstruction entirely and hit the same problem.

Each entry of the rewritten INSERT targetlist contains a Var that
references the SELECT column that's assigned to that table column, and the
position of that SELECT column matches the position in the DuckDB result.
So now we map the DuckDB columns through those Vars instead of
sequentially, and we always reconstruct the targetlist. Entries without
such a Var are used directly: those are either default values added by the
rewriter, or unknown-type constants like `SELECT i, 'foo'` that the parser
inlines into the INSERT targetlist instead of creating a Var (which is
also why the previous code checked `pgduckdb_is_not_default_expr`, that
check is not needed anymore).

Finally, tables with dropped columns were not handled: the reconstructed
targetlist skipped them, but `ExecCheckPlanOutput` requires a NULL
constant at their position, so such inserts failed with "Query provides a
value for a dropped column".

JelteF committed:
Remove EXPLAIN whose output depends on the DuckDB build type

The `EXPLAIN INSERT ... JOIN` test included the DuckDB execution plan in
its output, but Debug builds of DuckDB keep a few extra PROJECTION nodes
in the plan that Release builds optimize away. So the test could never
pass for both build types, and CI runs the test suite with both. The
actual INSERT right after it still verifies that the query works.

JelteF committed:
Fix CI

JelteF committed:
Document that INSERT ... SELECT needs casts on r['col'] references

Assigning a `duckdb.unresolved_type` expression to a table column fails
during parse analysis, because Postgres only knows the actual type once
DuckDB executes the query. The existing casts from
`duckdb.unresolved_type` are explicit-only, so users have to write them
out in the SELECT list of an INSERT. This documents that in the
`duckdb.row` section of the type documentation, since that's where users
will run into it.

JelteF committed:
Don't log NOTICEs for INSERTs that DuckDB cannot execute

`IsAllowedPostgresInsert` logged its reasons for not using DuckDB at
NOTICE level whenever `duckdb.log_pg_explain` was enabled. That GUC is
meant for logging the Postgres EXPLAIN plans of scans that DuckDB does on
Postgres tables, so this abused it as a general verbosity channel. It was
also quite noisy: every plain `INSERT ... VALUES` statement caused a
"DuckDB does not support INSERT without a subquery" NOTICE, which polluted
the expected output of several unrelated regression tests.

Now these messages use the same DEBUG4 level as the other reasons in
`IsAllowedStatement` for not using DuckDB. This also means the expected
output of `query_filter` and `scan_postgres_tables` is identical to main
again.

JelteF committed:
Test that unsupported modifications throw an error when DuckDB is required

Now that the reasons for not using DuckDB are only logged at DEBUG4 level,
the test cases for unsupported modifications of Postgres tables (UPDATE,
DELETE, plain VALUES inserts, unsupported column types, and VALUES RTEs in
the subquery) didn't show anything anymore: those statements silently fell
back to Postgres execution. That fallback behaviour is still tested, but
now each case also has a variant that references `duckdb.query()`. Such
statements require DuckDB execution, so falling back to Postgres is not
possible and the user gets the actual error explaining why DuckDB cannot
execute them.

JelteF committed:
Allow VALUES in the INSERT subquery when DuckDB execution is required

INSERTs whose SELECT source contained a VALUES RTE were always rejected,
because literal input may differ between Postgres and DuckDB (e.g. bytea
representation and numeric rounding). That's a good reason to prefer
Postgres execution when we have the choice, but when the query requires
DuckDB anyway there is no fallback, so rejecting the query only forces
users to rewrite it. Something like this now simply works:

    INSERT INTO tbl
    SELECT r['a']::int, v.column2
    FROM read_csv('data.csv') r
    JOIN (VALUES (2, 'yoyo')) v ON r['a']::int = v.column1;

The VALUES check now only applies when `throw_error` is false, since
that's only the case when DuckDB execution is optional. The planner
internal callers of `IsAllowedPostgresInsert` now pass throw_error=true,
because by the time they run the decision to use DuckDB has already been
made. For that to be safe the target-table check had to move before the
commandType check, so that UPDATE/DELETE statements on DuckDB tables
still return false instead of erroring when `DuckdbPrepare` translates
them.

JelteF committed:
Simplify targetlist reconstruction for INSERT ... SELECT

`ReconstructTargetListForInsert` and `CoerceTargetList` were one logical
operation split across two passes in two different functions:
`CreatePlan` reordered the DuckDB scan its targetlist into table-column
order, and `DuckdbPlanNode` later walked the same list again to coerce
the types. The types that `CoerceTargetList` extracted from the original
INSERT subplan using `ExecTypeFromTL` are simply the types of the table
columns, which `ReconstructTargetListForInsert` already had directly
available in its `TupleDesc`. So now the coercion happens right where a
column of the DuckDB result is assigned to a table column. The entries
for unassigned columns never need a coercion: default expressions were
already coerced by the rewriter, and the NULL constants are created with
the column its own type.

The custom `FindSubqueryVarWalker` is also replaced by the standard
`pull_var_clause` helper. And the TargetEntry taken from the DuckDB
targetlist doesn't need to be copied before modifying it: each SELECT
column is assigned to at most one table column, and the old targetlist
is discarded wholesale after the reconstruction.

JelteF committed:
Fix compilation on PG19

PG19 added an `ExplainState *` parameter to `standard_planner`, so the
call for INSERTs into Postgres tables needs to pass an extra NULL there,
just like the existing call in `postgres_table_reader.cpp` does.

JelteF committed:
Fix outdated comment about varno 1 always being correct

Since INSERTs into Postgres tables reuse the plan built by
`standard_planner`, it's no longer true that the final plan always has
the CustomScan as its only RTE. The comment in `CreatePlan` now mentions
that exception and points to the varno fixup in `DuckdbPlanNode`.

JelteF committed:
Always fill in the CustomScan varno after building the final rtable

`CreatePlan` hardcoded varno 1 in the `custom_scan_tlist`, which was only
correct for plans where the CustomScan is the only RTE. For INSERTs into
Postgres tables the varnos were patched up afterwards, since there the RTE
gets appended to the rtable that `standard_planner` built. Passing the
right varno to `CreatePlan` instead is not possible, because the RTE's
final position is only known after `standard_planner` runs, which happens
after `CreatePlan` on purpose: whether the DuckDB plan can be created at
all determines whether we use DuckDB in the first place.

So instead `CreatePlan` now uses the invalid varno 0, and both code paths
in `DuckdbPlanNode` fill in the actual position with `SetCustomScanVarno`
once they've built the rtable. This way a forgotten fixup in some future
code path fails loudly, instead of silently reading whatever RTE is at
position 1.

JelteF committed:
Return directly instead of tracking a `ret` variable

The `ret = false; break;` dance in `IsAllowedPostgresInsert` existed so
that `RelationClose` always runs before returning. But the `elog` right
before it already skips `RelationClose` whenever `throw_error` is set,
which is fine because the transaction abort releases the reference on the
relation. That same reasoning allows simply closing the relation and
returning directly in the non-throwing case.

JelteF committed:
Remove unnecessary comment

JelteF committed:
Simplify code and improve comments

JelteF committed:
Use assignment casts when coercing DuckDB results for INSERT

The casts that align the types of the DuckDB result with the types of the
table columns were created with `COERCION_EXPLICIT`. Explicit casts have
different semantics than the assignment casts that Postgres uses when
executing an INSERT itself: most importantly they silently truncate values
that don't fit the column type. So inserting a 6 character string into a
`varchar(3)` column stored just the first 3 characters, where Postgres
execution throws "value too long for type character varying(3)". Using
`COERCION_ASSIGNMENT` makes DuckDB execution behave the same as Postgres
execution here.

JelteF committed:
Disallow CTEs and EXPLAIN ANALYZE for INSERTs into Postgres tables

Both of these were broken in ways that could easily go unnoticed.

Since only the SELECT source of the INSERT is deparsed and sent to DuckDB,
a CTE attached to the INSERT statement itself was lost in that deparse:

    WITH x AS (...) INSERT INTO tbl SELECT * FROM x;

The deparsed query was `SELECT * FROM x` without the WITH clause, which
usually resulted in a "table x does not exist" error, but could also
silently read from an unrelated DuckDB-visible table named `x`. CTEs
inside the SELECT itself keep working, those are part of the deparsed
query.

EXPLAIN ANALYZE of such an INSERT executed the SELECT inside a DuckDB-side
EXPLAIN ANALYZE, which returns no rows to the ModifyTable node. So the
statement reported `actual rows=0` and silently inserted nothing, while
EXPLAIN ANALYZE of a Postgres-executed INSERT (and of an INSERT into a
DuckDB table) actually inserts the rows. Until we support this properly we
throw an error instead of silently doing nothing.

JelteF committed:
Add tests for generated columns and prepared statements

These both already worked, but weren't covered by any test yet. Columns
that the INSERT doesn't assign get a NULL placeholder in the targetlist,
after which the ModifyTable node computes `GENERATED ALWAYS AS ... STORED`
and `serial` columns as usual. And prepared statements with parameters in
the WHERE clause of the SELECT work fine, including when they switch to a
generic plan after repeated execution. Parameters in the SELECT list
itself don't work, but that's a pre-existing limitation that plain SELECT
statements have too, because DuckDB cannot determine the result type of
the prepared query.

JelteF committed:
Support CTEs on INSERTs into Postgres tables

Instead of rejecting CTEs that are attached to the INSERT statement
itself, we now move them into the SELECT query before deparsing it, so
that they end up in the WITH clause of the query that DuckDB executes:

    WITH x AS (SELECT r['a']::int a FROM read_csv('data.csv') r)
    INSERT INTO tbl SELECT * FROM x;

Because the CTEs move down a query level, all references to them from
within the SELECT need their levelsup decremented, similar to what
`pull_up_simple_subquery` does. That happens before attaching the moved
CTEs to the SELECT, so that references between those CTEs themselves stay
unchanged. When the SELECT has CTEs of its own, both lists are combined
into a single WITH clause. The one thing that gets lost that way is
Postgres its shadowing semantics when the same name appears in both
lists, but luckily DuckDB throws a "Duplicate CTE name" error for that
instead of silently picking one of the two.

JelteF commented:
@YuweiXiao thanks for all your work on this. It took a while, but it's merged now!
